channeldb/channel.go

```diff
@@ -3216,6 +3216,280 @@ const (
 	Abandoned ClosureType = 5
 )
 
+// LocalForceCloseInfo encapsulates information that led to a channel being
+// force closed with the initiator being local. Currently local force closes
+// can be (1) purely user initiated, (2) automatic due to certain link errors,
+// or (3) automatic due to certain on-chain HTLC conditions. As such, only one
+// of the three fields in this struct will ever have meaningful information. But
+// this may be extended in the future. For example, we may decide to log HTLC
+// information that doesn't automatically result in a force close, but leaves
+// the channel in a state whereby the user has no option but to force close.
+type LocalForceCloseInfo struct {
+	// UserInitiated is true iff the force close was specifically initiated
+	// by the user.
+	UserInitiated bool
+
+	// LinkFailureError is a non-empty string iff the force close was
+	// automatically initiated following from a link failure.
+	LinkFailureError string
+
+	// HtlcActions is a non-empty map iff the force close was automatically
+	// initiated by an on-chain trigger such as HTLC timeout.
+	HtlcActions map[string][]HTLC
+}
+
+const (
+
+	// userInitiatedType is used to serialize/deserialize UserInitiated
+	// field in LocalForceCloseInfo.
+	userInitiatedType tlv.Type = 0
+
+	// linkFailureErrorType is used to serialize/deserialize
+	// linkFailureError field in LocalForceCloseInfo.
+	linkFailureErrorType tlv.Type = 1
+
+	// htlcActionsType is used to serialize/deserialize HtlcActions field
+	// in LocalForceCloseInfo.
+	htlcActionsType tlv.Type = 2
+
+	// ActionKeyType is used to serialize/deserialize the action key in the
+	// htlcActions map in LocalForceCloseInfo.
+	ActionKeyType tlv.Type = 3
+)
+
+// HtlcActionRecordSize returns the amount of bytes this TLV record will occupy
+// when encoded.
+func HtlcActionRecordSize(a *map[string][]HTLC) func() uint64 {
+	var (
+		b   bytes.Buffer
+		buf [8]byte
+	)
+
+	// We know that encoding works since the tests pass in the build this
+	// file is checked into, so we'll simplify things and simply encode it
+	// ourselves then report the total amount of bytes used.
+	if err := HtlcActionEncoder(&b, a, &buf); err != nil {
+		// This should never error out, but we log it just in case it
+		// does.
+		log.Errorf("encoding the amp invoice state failed: %v", err)
+	}
+
+	return func() uint64 {
+		return uint64(len(b.Bytes()))
+	}
+}
+
+// HtlcActionEncoder is a custom TLV encoder for the htlcAction record.
+func HtlcActionEncoder(w io.Writer, val interface{}, buf *[8]byte) error {
```

Comment:
This would be great, though we currently have no fuzz tests for `channeldb`.  Writing those tests is probably a bigger project to tackle separately, rather than holding up this PR.

See https://github.com/lightningnetwork/lnd/pull/8045 for some progress on `channeldb` fuzzing.